File: product-of-others/productOfOthers.java
import java.util.Arrays;

public class ProductOfOthers {

  public static void main(String[] args) {
    int[] numList = new int[]{2,3,4,5};

    System.out.println(Arrays.toString(productOfOthers(numList)));
  }

  public static int[] productOfOthers(int[] numList) {
    int[] productsBeforeIndex = new int[numList.length];
    int[] productsExceptAtIndex = new int[numList.length];

    int currentProductBefore = 1;

    // Generate array of product of numbers before index i
    for (int i=0; i<numList.length; i++) {
      productsBeforeIndex[i] = currentProductBefore;
      currentProductBefore *= numList[i];
    }

    int currentProductAfter = 1;

    // for each index, multiply the product before the current index
    // times the product after the index
    // Keep track of currentProductAfter to avoid having to create
    // another array for storing productsAfterIndex
    for (int i=numList.length-1; i>=0; i--) {
      productsExceptAtIndex[i] = productsBeforeIndex[i] * currentProductAfter;
      currentProductAfter *= numList[i];
    }

    return productsExceptAtIndex;
  }

}

File: product-of-others/product_of_others.py
def product_of_others(num_list):
  products_before_index = []
  products_except_at_index = [None] * len(num_list)

  current_product_before = 1
  current_product_after = 1

  for num in num_list:
    products_before_index.append(current_product_before)
    current_product_before *= num

  i = len(num_list) - 1
  while i >= 0:
    products_except_at_index[i] = products_before_index[i] * current_product_after
    current_product_after *= num_list[i]
    i -= 1

  return products_except_at_index

if __name__ == "__main__":
  num_list = [2,3,4,5]

  print(product_of_others(num_list))